File: OpenStreetMap/iOS/GpxViewController.m
//
//  GpxViewController.m
//  Go Map!!
//
//  Created by Bryce on 2/26/13.
//  Copyright (c) 2013 Bryce. All rights reserved.
//

#import <MessageUI/MessageUI.h>

#import "AppDelegate.h"
#import "DLog.h"
#import "MapView.h"
#import "GpxConfigureViewController.h"
#import "GpxLayer.h"
#import "GpxViewController.h"
#import "OsmMapData.h"



@interface GpxTrackTableCell : UITableViewCell <UIActionSheetDelegate, MFMailComposeViewControllerDelegate>
@property (assign,nonatomic)	IBOutlet	UILabel				*	startDate;
@property (assign,nonatomic)	IBOutlet	UILabel				*	duration;
@property (assign,nonatomic)	IBOutlet	UILabel				*	details;
@property (strong,nonatomic)				GpxTrack			*	gpxTrack;
@property (assign,nonatomic)				GpxViewController	*	tableView;
@end
@implementation GpxTrackTableCell
-(IBAction)doAction:(id)sender
{
	UIActionSheet * sheet = [[UIActionSheet alloc] initWithTitle:NSLocalizedString(@"Share",nil) delegate:self cancelButtonTitle:NSLocalizedString(@"Cancel",nil) destructiveButtonTitle:nil otherButtonTitles:NSLocalizedString(@"Upload to OSM",nil), NSLocalizedString(@"Mail",nil), nil];
	[sheet showInView:self];
}
- (void)actionSheet:(UIActionSheet *)actionSheet didDismissWithButtonIndex:(NSInteger)buttonIndex
{
	if ( buttonIndex == actionSheet.cancelButtonIndex ) {
		return;
	}
	if ( buttonIndex == 0 ) {
		// upload
		[self.tableView shareTrack:_gpxTrack];
	} else if ( buttonIndex == 1 ) {
		// mail
		if ( [MFMailComposeViewController canSendMail] ) {
			AppDelegate * appDelegate = [AppDelegate getAppDelegate];
			MFMailComposeViewController * mail = [[MFMailComposeViewController alloc] init];
			mail.mailComposeDelegate = self;
			[mail setSubject:[NSString stringWithFormat:@"%@ GPX file: %@", appDelegate.appName, self.gpxTrack.creationDate]];
			[mail addAttachmentData:self.gpxTrack.gpxXmlData mimeType:@"application/gpx+xml" fileName:[NSString stringWithFormat:@"%@.gpx", self.gpxTrack.creationDate]];
			[self.tableView presentViewController:mail animated:YES completion:nil];
		} else {
			UIAlertView * alert = [[UIAlertView alloc] initWithTitle:NSLocalizedString(@"Cannot compose message",nil) message:NSLocalizedString(@"Mail delivery is not available on this device",nil) delegate:nil cancelButtonTitle:NSLocalizedString(@"OK",nil) otherButtonTitles:nil];
			[alert show];
		}
	}
}
-(void)mailComposeController:(MFMailComposeViewController*)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError*)error
{
	[controller dismissViewControllerAnimated:YES completion:nil];
}
@end


@interface GpxTrackExpirationCell : UITableViewCell
@property (assign,nonatomic)	IBOutlet	UIButton *	expirationButton;
@end
@implementation GpxTrackExpirationCell
@end


@implementation GpxViewController

- (void)viewDidLoad
{
	[super viewDidLoad];

	_navigationBar.topItem.rightBarButtonItem = self.editButtonItem;
	self.navigationItem.rightBarButtonItem = self.editButtonItem;

	AppDelegate * appDelegate = [AppDelegate getAppDelegate];
	[appDelegate.mapView.gpxLayer loadTracksInBackgroundWithProgress:^{
		[self.tableView reloadData];
	}];
}


-(void)viewWillAppear:(BOOL)animated
{
	[super viewWillAppear:animated];

	AppDelegate * appDelegate = [AppDelegate getAppDelegate];
	if ( appDelegate.mapView.gpxLayer.activeTrack ) {
		[self startTimerForStartDate:appDelegate.mapView.gpxLayer.activeTrack.creationDate];
	}
}

-(void)viewWillDisappear:(BOOL)animated
{
	[super viewWillDisappear:animated];
	
	[_timer invalidate];
	_timer = nil;
}

-(void)startTimerForStartDate:(NSDate *)date
{
	NSDate * now = [NSDate date];
	NSTimeInterval delta = [now timeIntervalSinceDate:date];
	delta = 1 - fmod(delta,1.0);
	date = [now dateByAddingTimeInterval:delta];
	_timer = [[NSTimer alloc] initWithFireDate:date interval:1.0 target:self selector:@selector(timerFired:) userInfo:nil repeats:YES];

	NSRunLoop *runloop = [NSRunLoop currentRunLoop];
	[runloop addTimer:_timer forMode:NSDefaultRunLoopMode];
}

-(void)timerFired:(NSTimer *)timer
{
	AppDelegate * appDelegate = [AppDelegate getAppDelegate];
	GpxTrack * track = appDelegate.mapView.gpxLayer.activeTrack;
	if ( track ) {
		NSIndexPath * index = [NSIndexPath indexPathForRow:0 inSection:0];
		[self.tableView reloadRowsAtIndexPaths:@[index] withRowAnimation:UITableViewRowAnimationNone];
	} else {
		[_timer invalidate];
		_timer = nil;
	}
}

-(IBAction)cancel:(id)sender
{
	[self dismissViewControllerAnimated:YES completion:nil];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
	return 3;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
	if ( section == 0 ) {
		return 1;
	} else if ( section == 1 ) {
		AppDelegate * appDelegate = [AppDelegate getAppDelegate];
		return appDelegate.mapView.gpxLayer.previousTracks.count;
	} else if ( section == 2 ) {
		return 1;
	} else {
		return 0;
	}
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
	switch (section) {
		case 0:
			return @"Current Track";
		case 1:
			return @"Previous Tracks";
		case 2:
			return @"Configure";
		default:
			return nil;
	}
}

- (NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section
{
	if ( section == 0 ) {
		return @"A GPX Track records your path as you travel along a road or trail";
	}
	return nil;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
	AppDelegate * appDelegate = [AppDelegate getAppDelegate];

	if ( indexPath.section == 0 && appDelegate.mapView.gpxLayer.activeTrack == nil ) {
		// no active track
		UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:nil];
		cell.textLabel.text = @"No active track";
		return cell;
	}
	if ( indexPath.section == 2 ) {
		GpxTrackExpirationCell * cell = [tableView dequeueReusableCellWithIdentifier:@"GpxTrackExpirationCell" forIndexPath:indexPath];
		NSNumber * expirationDays = [[NSUserDefaults standardUserDefaults] objectForKey:USER_DEFAULTS_GPX_EXPIRATIION_KEY];
		NSInteger expiration = [expirationDays integerValue];
		NSString * title = expiration <= 0 ? @"Never" : [NSString stringWithFormat:@"%ld Days",(long)expiration];
		[cell.expirationButton setTitle:title forState:UIControlStateNormal];
		[cell.expirationButton sizeToFit];
		return cell;
	}

	NSArray * prevTracks = appDelegate.mapView.gpxLayer.previousTracks;
	GpxTrack *	track = indexPath.section == 0 ? appDelegate.mapView.gpxLayer.activeTrack : prevTracks[ prevTracks.count - indexPath.row - 1 ];
	NSInteger	dur = (NSInteger)round(track.duration);
	NSString * startDate = [NSDateFormatter localizedStringFromDate:track.creationDate dateStyle:NSDateFormatterShortStyle timeStyle:NSDateFormatterShortStyle];
	NSString * duration = [NSString stringWithFormat:@"%d:%02d:%02d", (int)(dur/3600), (int)(dur/60%60), (int)(dur%60)];
	NSString * meters = [NSString stringWithFormat:@"%ld meters, %ld points", (long)track.distance, (long)track.points.count];
	GpxTrackTableCell * cell = [tableView dequeueReusableCellWithIdentifier:@"GpxTrackTableCell" forIndexPath:indexPath];
	cell.startDate.text = startDate;
	cell.duration.text = duration;
	cell.details.text = meters;
	cell.gpxTrack = track;
	cell.tableView = self;
	return cell;
}

// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    return indexPath.section == 1;
}

// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
		AppDelegate * appDelegate = [AppDelegate getAppDelegate];
		GpxTrack * track = appDelegate.mapView.gpxLayer.previousTracks[ indexPath.row ];
		[appDelegate.mapView.gpxLayer deleteTrack:track];

		[tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}


/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
{
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

#pragma mark - Table view delegate

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	if ( indexPath.section == 0 ) {
		// active track
	} else if ( indexPath.section == 2 ) {
		// configuration
	} else if ( indexPath.section == 1 ) {
		AppDelegate * appDelegate = [AppDelegate getAppDelegate];
		GpxTrack * track = appDelegate.mapView.gpxLayer.previousTracks[ indexPath.row ];
		[appDelegate.mapView.gpxLayer centerOnTrack:track];
		[self.navigationController popToRootViewControllerAnimated:YES];
	}
}

-(void)shareTrack:(GpxTrack *)track
{
	UIAlertView * progress = [[UIAlertView alloc] initWithTitle:@"Uploading GPX..." message:@"Please wait" delegate:self cancelButtonTitle:nil otherButtonTitles:nil];
	[progress show];

	// let progress window display before we submit work
	dispatch_async(dispatch_get_main_queue(), ^{
		AppDelegate * appDelegate = [AppDelegate getAppDelegate];

		NSString * url = [OSM_API_URL stringByAppendingFormat:@"api/0.6/gpx/create"];

		NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:url]];
		NSString * boundary = @"----------------------------d10f7aa230e8";
		[request setHTTPMethod:@"POST"];
		NSString * contentType = [NSString stringWithFormat:@"multipart/form-data; boundary=%@",boundary];
		[request setValue:contentType	forHTTPHeaderField:@"Content-Type"];
		[request setValue:@"close"		forHTTPHeaderField:@"Connection"];

		NSDateFormatter * dateFormatter = [NSDateFormatter new];
		[dateFormatter setDateFormat:@"yyyy_MM_dd__HH_mm_ss"];
		NSString * startDateFile = [dateFormatter stringFromDate:track.creationDate];
		NSString * startDateFriendly = [NSDateFormatter localizedStringFromDate:track.creationDate dateStyle:NSDateFormatterShortStyle timeStyle:NSDateFormatterShortStyle];

		NSMutableData * body = [NSMutableData new];
		[body appendData:[[NSString stringWithFormat:@"--%@\r\n",boundary]   dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[[NSString stringWithString:[NSString stringWithFormat: @"Content-Disposition: form-data; name=\"file\"; filename=\"GoMap__%@.gpx\"\r\n",startDateFile]] dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[@"Content-Type: application/octet-stream\r\n\r\n" dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:track.gpxXmlData];

		[body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n",boundary]   dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[[NSString stringWithString:[NSString stringWithFormat: @"Content-Disposition: form-data; name=\"description\"\r\n\r\n"]] dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[[NSString stringWithFormat:@"Go Map!! %@",startDateFriendly] dataUsingEncoding:NSUTF8StringEncoding]];

		[body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n",boundary]   dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[[NSString stringWithString:[NSString stringWithFormat: @"Content-Disposition: form-data; name=\"tags\"\r\n\r\n"]] dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[@"GoMap" dataUsingEncoding:NSUTF8StringEncoding]];

		[body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n",boundary]   dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[[NSString stringWithString:[NSString stringWithFormat: @"Content-Disposition: form-data; name=\"public\"\r\n\r\n"]] dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[@"1" dataUsingEncoding:NSUTF8StringEncoding]];

		[body appendData:[[NSString stringWithFormat:@"\r\n--%@\r\n",boundary]   dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[[NSString stringWithString:[NSString stringWithFormat: @"Content-Disposition: form-data; name=\"visibility\"\r\n\r\n"]] dataUsingEncoding:NSUTF8StringEncoding]];
		[body appendData:[@"public" dataUsingEncoding:NSUTF8StringEncoding]];

		[body appendData:[[NSString stringWithFormat:@"\r\n--%@--\r\n",boundary] dataUsingEncoding:NSUTF8StringEncoding]];
		[request setHTTPBody:body];

		[request setValue:[NSString stringWithFormat:@"%ld", (long)body.length] forHTTPHeaderField:@"Content-Length"];

		NSString * auth = [NSString stringWithFormat:@"%@:%@", appDelegate.userName, appDelegate.userPassword];
		auth = [OsmMapData encodeBase64:auth];
		auth = [NSString stringWithFormat:@"Basic %@", auth];
		[request setValue:auth forHTTPHeaderField:@"Authorization"];

//		DLog(@"body = %@",[NSString stringWithUTF8String:body.bytes] );

		[NSURLConnection sendAsynchronousRequest:request queue:[NSOperationQueue mainQueue] completionHandler:^(NSURLResponse * response, NSData * data, NSError * error) {
			[progress dismissWithClickedButtonIndex:0 animated:YES];

			NSHTTPURLResponse * httpResponse = [response isKindOfClass:[NSHTTPURLResponse class]] ? (id)response : nil;
			if ( httpResponse.statusCode == 200 ) {
				// ok
				UIAlertView * success = [[UIAlertView alloc] initWithTitle:@"GPX Upload Complete" message:nil delegate:self cancelButtonTitle:@"OK" otherButtonTitles:nil];
				[success show];
			} else {
				DLog(@"response = %@\n",response);
				DLog(@"data = %s", (char *)data.bytes);
				NSString * errorMessage = nil;
				if ( data.length > 0 ) {
					errorMessage = [[NSString alloc] initWithBytes:data.bytes length:data.length encoding:NSUTF8StringEncoding];
				} else {
					errorMessage = error.localizedDescription;
				}
				// failure
				UIAlertView * failure = [[UIAlertView alloc] initWithTitle:@"GPX Upload Failed" message:errorMessage delegate:self cancelButtonTitle:@"OK" otherButtonTitles:nil];
				[failure show];
			}
		}];
	});
}


#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
	if ( [segue.destinationViewController isKindOfClass:[GpxConfigureViewController class]] ) {
		GpxConfigureViewController * dest = segue.destinationViewController;
		dest.expirationValue = [[NSUserDefaults standardUserDefaults] objectForKey:USER_DEFAULTS_GPX_EXPIRATIION_KEY];
		dest.completion = ^(NSNumber * pick){
			[[NSUserDefaults standardUserDefaults] setObject:pick forKey:USER_DEFAULTS_GPX_EXPIRATIION_KEY];

			if ( pick.doubleValue > 0 ) {
				AppDelegate * appDelegate = [AppDelegate getAppDelegate];
				NSDate * cutoff = [NSDate dateWithTimeIntervalSinceNow:-pick.doubleValue*24*60*60];
				[appDelegate.mapView.gpxLayer trimTracksOlderThan:cutoff];
			}
			[self.tableView reloadData];
		};
	}
}

@end

File: src/iOS/DisplaySettings/GpxViewController.swift
//
//  GpxViewController.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 2/26/13.
//  Copyright (c) 2013 Bryce Cogswell. All rights reserved.
//

import UIKit

class GpxTrackTableCell: UITableViewCell, UIActionSheetDelegate {
	@IBOutlet var startDate: UILabel!
	@IBOutlet var duration: UILabel!
	@IBOutlet var details: UILabel!
	@IBOutlet var uploadButton: UIButton!
	var gpxTrack = GpxTrack()
	var tableView: GpxViewController?

	@IBAction func doAction(_ sender: Any) {
		let alert = UIAlertController(
			title: NSLocalizedString("Share", comment: "Title for sharing options"),
			message: nil,
			preferredStyle: .actionSheet)
		alert.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: ""), style: .cancel, handler: nil))
		alert.addAction(UIAlertAction(title: NSLocalizedString("Upload to OSM", comment: ""),
		                              style: .default,
		                              handler: { _ in
		                              	self.tableView?.share(self.gpxTrack)
		                              }))
		alert.addAction(UIAlertAction(
			title: NSLocalizedString("Share", comment: "Open iOS sharing sheet"),
			style: .default, handler: { _ in
				let creationDate = self.gpxTrack.creationDate
				let appName = AppDelegate.shared.appName()
				let fileName = "\(appName) \(creationDate).gpx"

				// get a copy of GPX as a string
				guard let gpx = self.gpxTrack.gpxXmlString() else { return }

				// create a temporary file and write the gpx to it
				let fileUrl = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)
				try? FileManager.default.removeItem(at: fileUrl)
				guard (try? gpx.write(to: fileUrl, atomically: true, encoding: .utf8)) != nil else {
					return
				}

				// Create the share activity
				let controller = UIActivityViewController(activityItems: [fileName, fileUrl] as [Any],
				                                          applicationActivities: nil)
				controller.completionWithItemsHandler = { _, completed, _, _ in
					if completed {
						let gpxLayer = AppDelegate.shared.mapView.gpxLayer
						gpxLayer.markTrackUploaded(self.gpxTrack)
						self.tableView?.tableView.reloadData()
					}
				}

				// Preset the share sheet to the user
				self.tableView?.present(controller, animated: true)
			}))

		// set location of popup
		let button = sender as? UIButton
		alert.popoverPresentationController?.sourceView = button
		alert.popoverPresentationController?.sourceRect = button?.bounds ?? CGRect.zero

		tableView?.present(alert, animated: true)
	}
}

class GpxTrackBackgroundCollection: UITableViewCell {
	@IBOutlet var enableBackground: UISwitch!

	@IBAction func enableBackground(_ sender: Any) {
		let toggle = sender as? UISwitch
		let appDelegate = AppDelegate.shared
		appDelegate.mapView.gpsInBackground = toggle?.isOn ?? false
	}
}

class GpxTrackImportHealthKit: UITableViewCell {
	@IBOutlet var importButton: UIButton!
	@IBOutlet var activityIndicator: UIActivityIndicatorView!
	var vc: UIViewController!

	@IBAction func importHealthKitRoutes(_ sender: Any) {
		activityIndicator.isHidden = false
		activityIndicator.startAnimating()
		HealthKitRoutes.shared.getWorkoutRoutes { result in
			self.activityIndicator.stopAnimating()
			self.activityIndicator.isHidden = true
			switch result {
			case let .failure(error):
				let alert = UIAlertController(title: NSLocalizedString("HealthKit", comment: "iOS HealthKit framework"),
				                              message: error.localizedDescription,
				                              preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
												style: .default,
												handler: nil))
				self.vc.present(alert, animated: true)
			case let .success(routes):
				let origCount = AppDelegate.shared.mapView.gpxLayer.previousTracks.count
				for route in routes {
					let gpx = GpxTrack()
					gpx.name = "HealthKit"
					for point in route {
						gpx.addPoint(point)
					}
					gpx.finish()
					AppDelegate.shared.mapView.gpxLayer.addGPX(track: gpx, center: false)
				}
				if let tableView = self.vc.view as? UITableView {
					tableView.reloadData()
				}
				let newCount = AppDelegate.shared.mapView.gpxLayer.previousTracks.count - origCount
				let message = NSLocalizedString(
					"\(newCount) new route(s) imported,\n\(routes.count - newCount) duplicate(s) skipped",
					comment: "")
				let alert = UIAlertController(title: NSLocalizedString("HealthKit", comment: "iOS HealthKit framework"),
				                              message: message,
				                              preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
												style: .default,
												handler: nil))
				self.vc.present(alert, animated: true)
			}
		}
	}
}

class GpxTrackExpirationCell: UITableViewCell {
	@IBOutlet var expirationButton: UIButton!
}

class GpxViewController: UITableViewController {
	private var timer: Timer?
	@IBOutlet var navigationBar: UINavigationBar!

	@IBAction func cancel(_ sender: Any) {
		dismiss(animated: true)
	}

	func share(_ track: GpxTrack) {
		let progress = UIAlertController(
			title: NSLocalizedString("Uploading GPX...", comment: ""),
			message: NSLocalizedString("Please wait", comment: ""),
			preferredStyle: .alert)
		present(progress, animated: true)

		// let progress window display before we submit work
		DispatchQueue.main.async(execute: {
			let url = OSM_API_URL + "api/0.6/gpx/create"

			guard var request = AppDelegate.shared.oAuth2.urlRequest(string: url) else { return }
			let boundary = "----------------------------d10f7aa230e8"
			request.httpMethod = "POST"
			let contentType = "multipart/form-data; boundary=\(boundary)"
			request.setValue(contentType, forHTTPHeaderField: "Content-Type")
			request.setValue("close", forHTTPHeaderField: "Connection")

			let dateFormatter = DateFormatter()
			dateFormatter.dateFormat = "yyyy_MM_dd__HH_mm_ss"
			let startDateFile = dateFormatter.string(from: track.creationDate)
			let startDateFriendly = DateFormatter.localizedString(
				from: track.creationDate,
				dateStyle: .short,
				timeStyle: .short)

			var body = ""
			body += "--\(boundary)\r\n"
			body += "Content-Disposition: form-data; name=\"file\"; filename=\"GoMap__\(startDateFile).gpx\"\r\n"
			body += "Content-Type: application/octet-stream\r\n\r\n"
			body += track.gpxXmlString()!
			body += "\r\n--\(boundary)\r\n"
			body += "Content-Disposition: form-data; name=\"description\"\r\n\r\n"
			body += "Go Map!! \(startDateFriendly)"
			body += "\r\n--\(boundary)\r\n"
			body += "Content-Disposition: form-data; name=\"tags\"\r\n\r\n"
			body += "GoMap"
			body += "\r\n--\(boundary)\r\n"
			body += "Content-Disposition: form-data; name=\"public\"\r\n\r\n"
			body += "1"
			body += "\r\n--\(boundary)\r\n"
			body += "Content-Disposition: form-data; name=\"visibility\"\r\n\r\n"
			body += "public"
			body += "\r\n--\(boundary)--\r\n"
			request.httpBody = body.data(using: .utf8)
			request.setValue(String(format: "%ld", body.count), forHTTPHeaderField: "Content-Length")

			URLSession.shared.data(with: request, completionHandler: { result in
				DispatchQueue.main.async(execute: {
					progress.dismiss(animated: true)

					switch result {
					case .success:
						let success = UIAlertController(
							title: NSLocalizedString("GPX Upload Complete", comment: ""),
							message: nil,
							preferredStyle: .alert)
						success.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
						                                style: .cancel,
						                                handler: nil))
						self.present(success, animated: true)

						// mark track as uploaded in UI
						let gpxLayer = AppDelegate.shared.mapView.gpxLayer
						gpxLayer.markTrackUploaded(track)
						self.tableView?.reloadData()
					case let .failure(error):
						let failure = UIAlertController(
							title: NSLocalizedString("GPX Upload Failed", comment: ""),
							message: "\(error.localizedDescription)",
							preferredStyle: .alert)
						failure.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
						                                style: .cancel,
						                                handler: nil))
						self.present(failure, animated: true)
					}
				})
			})
		})
	}

	override func viewDidLoad() {
		super.viewDidLoad()

		tableView?.estimatedRowHeight = 44
		tableView?.rowHeight = UITableView.automaticDimension

		navigationItem.rightBarButtonItem = editButtonItem

		AppDelegate.shared.mapView.gpxLayer.loadTracksInBackground(withProgress: {
			self.tableView?.reloadData()
		})
	}

	override func viewDidAppear(_ animated: Bool) {
		super.viewDidAppear(animated)

		if let track = AppDelegate.shared.mapView.gpxLayer.activeTrack {
			startTimer(forStart: track.creationDate)
		}
	}

	override func viewWillDisappear(_ animated: Bool) {
		super.viewWillDisappear(animated)
		timer?.invalidate()
		timer = nil
	}

	func startTimer(forStart date: Date) {
		let now = Date()
		var delta = now.timeIntervalSince(date)
		delta = 1 - fmod(delta, 1.0)
		let date = now.addingTimeInterval(delta)
		timer = Timer(fire: date, interval: 1.0, repeats: true, block: { [weak self] timer in
			guard let self = self else { return }
			if AppDelegate.shared.mapView.gpxLayer.activeTrack != nil {
				let index = IndexPath(row: 0, section: SECTION_ACTIVE_TRACK)
				self.tableView?.reloadRows(at: [index], with: .none)
			} else {
				timer.invalidate()
				self.timer = nil
			}
		})
		RunLoop.current.add(timer!, forMode: .default)
	}

	// MARK: - Table view data source

	override func numberOfSections(in tableView: UITableView) -> Int {
		return 3
	}

	override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		if section == SECTION_ACTIVE_TRACK {
			// active track
			return 1
		} else if section == SECTION_PREVIOUS_TRACKS {
			// previous tracks
			return AppDelegate.shared.mapView.gpxLayer.previousTracks.count
		} else if section == SECTION_CONFIGURE {
			// configuration
			return 3
		} else {
			return 0
		}
	}

	override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
		switch section {
		case SECTION_ACTIVE_TRACK:
			return NSLocalizedString("Current Track", comment: "current GPX track")
		case SECTION_PREVIOUS_TRACKS:
			return NSLocalizedString("Previous Tracks", comment: "previous GPX track")
		case SECTION_CONFIGURE:
			return NSLocalizedString("Configure", comment: "")
		default:
			return nil
		}
	}

	override func tableView(_ tableView: UITableView, titleForFooterInSection section: Int) -> String? {
		if section == SECTION_ACTIVE_TRACK {
			return NSLocalizedString("A GPX Track records your path as you travel along a road or trail", comment: "")
		}
		return nil
	}

	override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let mapView = AppDelegate.shared.mapView!
		let gpxLayer = mapView.gpxLayer

		if indexPath.section == SECTION_ACTIVE_TRACK, gpxLayer.activeTrack == nil {
			// no active track
			let cell = UITableViewCell(style: .default, reuseIdentifier: nil)
			cell.textLabel?.text = NSLocalizedString("No active track", comment: "GPX track")
			return cell
		}
		if indexPath.section == SECTION_CONFIGURE {
			// configuration section
			switch indexPath.row {
			case 0:
				// days before deleting
				let cell = tableView.dequeueReusableCell(
					withIdentifier: "GpxTrackExpirationCell",
					for: indexPath) as! GpxTrackExpirationCell
				let expirationDays = GpxLayer.expirationDays
				let expiration = expirationDays
				let title = expiration <= 0 ? NSLocalizedString("Never", comment: "Never delete old tracks") : String
					.localizedStringWithFormat(
						NSLocalizedString("%ld Days", comment: "One or more days"),
						expiration)
				cell.expirationButton.setTitle(title, for: .normal)
				cell.expirationButton.sizeToFit()
				return cell
			case 1:
				// enable background use
				let cell = tableView.dequeueReusableCell(
					withIdentifier: "GpxTrackBackgroundCollection",
					for: indexPath) as! GpxTrackBackgroundCollection
				cell.enableBackground.isOn = mapView.gpsInBackground
				return cell
			case 2:
				// HealthKit support
				let cell = tableView.dequeueReusableCell(
					withIdentifier: "GpxTrackImportHealthKit",
					for: indexPath) as! GpxTrackImportHealthKit
				cell.activityIndicator.isHidden = true
				cell.vc = self
				return cell
			default:
				assertionFailure()
			}
		}

		// active track or previous tracks
		let track = indexPath.section == SECTION_ACTIVE_TRACK ? gpxLayer.activeTrack!
			: gpxLayer.previousTracks[indexPath.row]
		let startDate = DateFormatter.localizedString(from: track.creationDate, dateStyle: .short, timeStyle: .short)
		let dur = Int(round(track.duration()))
		let duration = String(format: "%d:%02d:%02d", dur / 3600, dur / 60 % 60, dur % 60)
		let subtitle: String
		if track.wayPoints.count > 0 {
			subtitle = String.localizedStringWithFormat(
				NSLocalizedString("%ld meters, %ld points, %ld waypoints", comment: "length of a gpx track"),
				Int(track.lengthInMeters()),
				track.points.count,
				track.wayPoints.count)
		} else {
			subtitle = String.localizedStringWithFormat(
				NSLocalizedString("%ld meters, %ld points", comment: "length of a gpx track"),
				Int(track.lengthInMeters()),
				track.points.count)
		}
		let cell = tableView.dequeueReusableCell(
			withIdentifier: "GpxTrackTableCell",
			for: indexPath) as! GpxTrackTableCell
		cell.startDate.text = startDate
		cell.duration.text = duration
		cell.details.text = subtitle
		cell.gpxTrack = track
		cell.tableView = self
		let name = track.name
		if gpxLayer.uploadedTracks[name] != nil {
			cell.uploadButton.setImage(nil, for: .normal)
			cell.uploadButton.setTitle("\u{2714}", for: .normal)
		} else {
			let image = UIImage(named: "702-share")
			cell.uploadButton.setImage(image, for: .normal)
			cell.uploadButton.setTitle(nil, for: .normal)
		}
		return cell
	}

	// Override to support conditional editing of the table view.
	override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
		return indexPath.section == SECTION_PREVIOUS_TRACKS
	}

	// Override to support editing the table view.
	override func tableView(
		_ tableView: UITableView,
		commit editingStyle: UITableViewCell.EditingStyle,
		forRowAt indexPath: IndexPath)
	{
		if editingStyle == .delete {
			// Delete the row from the data source
			let gpxLayer = AppDelegate.shared.mapView.gpxLayer
			let track = gpxLayer.previousTracks[indexPath.row]
			gpxLayer.delete(track)

			tableView.deleteRows(at: [indexPath], with: .fade)
		} else if editingStyle == .insert {
			// Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
		}
	}

	// MARK: - Table view delegate

	override func tableView(_ tableView: UITableView, willSelectRowAt indexPath: IndexPath) -> IndexPath? {
		if indexPath.section == SECTION_CONFIGURE {
			return nil
		}
		if indexPath.section == SECTION_ACTIVE_TRACK, AppDelegate.shared.mapView.gpxLayer.activeTrack == nil {
			// don't allow selecting the active track if there is none
			return nil
		}
		return indexPath
	}

	override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
		if indexPath.section == SECTION_ACTIVE_TRACK {
			// active track
			let gpxLayer = AppDelegate.shared.mapView.gpxLayer
			gpxLayer.selectedTrack = gpxLayer.activeTrack
			if let selectedTrack = gpxLayer.selectedTrack {
				gpxLayer.center(on: selectedTrack)
			}
			navigationController?.dismiss(animated: true)
		} else if indexPath.section == SECTION_CONFIGURE {
			// configuration
		} else if indexPath.section == SECTION_PREVIOUS_TRACKS {
			let gpxLayer = AppDelegate.shared.mapView.gpxLayer
			let track = gpxLayer.previousTracks[indexPath.row]
			gpxLayer.selectedTrack = track
			gpxLayer.center(on: track)
			AppDelegate.shared.mapView.displayGpxLogs = true
			navigationController?.dismiss(animated: true)
		}
	}

	// MARK: - Navigation

	// In a storyboard-based application, you will often want to do a little preparation before navigation
	override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
		if let dest = segue.destination as? GpxConfigureViewController {
			dest.expirationValue = GpxLayer.expirationDays
			dest.completion = { pick in
				GpxLayer.expirationDays = pick

				if pick > 0 {
					let cutoff = Date(timeIntervalSinceNow: TimeInterval(-pick * 24 * 60 * 60))
					AppDelegate.shared.mapView.gpxLayer.trimTracksOlderThan(cutoff)
				}
				self.tableView?.reloadData()
			}
		}
	}
}

let SECTION_CONFIGURE = 0
let SECTION_ACTIVE_TRACK = 1
let SECTION_PREVIOUS_TRACKS = 2
